docs/proposal/20200506-single-controller-multitenancy.md

```diff
@@ -0,0 +1,869 @@
+
+<a name="FR9">FR9.</a> CAPA MUST support clusterctl move scenarios.
+
+### Non-Functional
+
+<a name="NFR8">NFR8.</a> Each instance of CAPA SHOULD be able to support 200 clusters using role assumption.
+
+<a name="NFR8">NFR9.</a> CAPA MUST call STS APIs only when necessary to prevent rate limiting.
+
+<a name="NFR8">NFR10.</a> Unit tests MUST exist for all credential provider code.
+
+<a name="NFR8">NFR11.</a> e2e tests SHOULD exist for all credential provider code.
+
+<a name="NFR8">NFR12.</a> Credential provider code COULD be audited by security engineers.
+
+### Implementation Details/Notes/Constraints
+
+The current implementation of CAPA requests a new AWS EC2 and ELB service per cluster and per machine and stores these
+in fields on the ClusterScope struct. ClusterScopes are reference values which are created per-reconciliation:
+
+``` go
+type ClusterScope struct {
+  logr.Logger
+  client      client.Client
+  patchHelper *patch.Helper
+
+  AWSClients
+  Cluster    *clusterv1.Cluster
+  AWSCluster *infrav1.AWSCluster
+}
+```
+
+The field AWSClients holds the actual AWS service clients, and is defined like so:
+
+``` go
+type AWSClients struct {
+  EC2             ec2iface.EC2API
+  ELB             elbiface.ELBAPI
+  SecretsManager  secretsmanageriface.SecretsManagerAPI
+  ResourceTagging resourcegroupstaggingapiiface.ResourceGroupsTaggingAPIAPI
+}
+```
+
+The signatures for the functions which create these instances are as follows:
+
+``` go
+func NewClusterScope(params ClusterScopeParams) (*ClusterScope, error) {
+  ...
+  return &ClusterScope{
+    ...
+  }, nil
+}
+```
+
+#### Proposed Changes
+
+This proposal borrows heavily from the Service APIs specification.
+
+In the initial implementation, all new resources will be scoped at the cluster level, this is to enable delegation of
+AWS accounts whilst preventing privilege escalation as per [FR4](#FR4). Reasons for this are documented in the
+[alternatives](#alternatives) section.
+
+##### Cluster API Provider AWS v1alpha3 types
+
+<strong><em>Changed Resources</strong></em>
+* `AWSCluster`
+
+<strong><em>New Resources</strong></em>
+
+<em>Cluster scoped resources</em>
+
+* `AWSClusterStaticPrincipal` represents a static AWS tuple of credentials.
+* `AWSClusterRolePrincipal` represents an intent to assume an AWS role for cluster management.
+
+<em>Namespace scoped resources</em>
+
+* `AWSServiceAccountPrincipal` represents the use of a Kubernetes service account for access
+  to AWS using federated identity.
+
+<strong><em>Changes to AWSCluster</em></strong>
+
+A new field is added to the `AWSClusterSpec` to reference an principal. We intend to use `corev1.LocalObjectReference` in
+order to ensure that the only objects that can be references are either in the same namespace or are scoped to the
+entire cluster.
+
+``` go
+type  AWSClusterSpec  struct {
+  PrincipalRef corev1.LocalObjectReference `json:"principalRef,omitempty"`
```

Comment:
I assume this is also marked as `// +optional`?

Reply:
Should this field be a pointer?